### ISSUE TITLE
vector storable () instance broken in current releases.

### ISSUE THREAD

**Issue by lehins** [haskell/vector#258]
Creating a non-empty or empty storable vector of () is still a problem:

```
λ> mv <- MVS.new 0 :: IO (MVS.MVector RealWorld ()) *** Exception: divide by zero
λ> VS.empty :: VS.Vector () *** Exception: divide by zero 
```

This seems like a related ticket #167 that was fixed on master over two years ago in this commit 3e2bd9f, but [this release](https://github.com/haskell/vector/blob/v0.12.0.3/Data/Vector/Storable/Mutable.hs#L133) doesn't have it.

**Comment by cartazio** [comment]
V0.12 branch series is the most recent release.

Unit is semanticly uniformstive for strict array representations.

I’m going to edit the ticket title and your message to focus on the
constructive parts. This complaining element doesn’t help anyone :)

**Comment by cartazio** [comment]
i'll see what needs to be backported onto the 0.12 series vs what needs a new major version cut for the releases.

an 0.13 should happen in the next few months, but i've some other commitments i'm prioritizing to catchup on first

**Comment by cartazio** [comment]
wait... i thought v0.12.0.3 was released ... let me check

**Comment by cartazio** [comment]
@lehins derp, i see. thats an easy backport.

if you look at some closed tickets, i was actively asking for what patches to backport, because theres a lot to keep track of. 

@lehins could you compile a "patches to backport for a 0.12.0.4 release" so i can see which ones are api stable / semantics improving/preserving to consider? (that one slipped through the cracks i guess ;) )

**Comment by cartazio** [comment]
i'll have to look at this change, i'm not sure if its the right one for that underlying problem. i'll look later this week,

email/irc is the best way to get my attention if you think somethings been overlooked, :)

**Comment by lehins** [comment]
@cartazio I believe the biggest problem here is not in the fact that a single issue was fixed and overlooked for the release, but the fact that development is being done on master, while bug fixes are being cherry picked by a maintainer before a release.

For example I have no idea which changes got skipped, or which ones are already on master and I really don't have time to dig trough old tickets just to see which ones got released and which ones didn't, this should be decided upon merging a request.
>  could you compile a "patches to backport for a 0.12.0.4 release" so i can see which ones are api stable / semantics improving/preserving to consider? (that one slipped through the cracks i guess ;) )

Here is what I'd recommend, feel free to disagree and even close that ticket if you feel like the way things are done now is the correct way.

I'd personally recommend having a "future branch", meaning that backwards incompatible changes go into that branch, for example now it would be a branch `v13`, while the current major version stays on master and all releases are made from master.

This way bug fixes will not be overlooked, and if someone submits a PR with API breaking changes, it will be easy to ask that contributor to redirect the target of a PR to a the "future branch" (eg. v13)

Reasoning behind this is the fact that breaking changes are introduced very seldom, while various patches are quite common.

PS. I'd really appreciate if you wouldn't change any content that I write. Fee free to close any ticket I open, but the semantics of the content I write I would really-really-really prefer to stay intact (modulo sum minor fixes of course). For example you changed the title, but the new title is totally orthogonal of what I was getting at. It's not the code, but the process that is broken. Sorry you felt like it had a "complaining element", I wrote it the way I did to emphasize the fact that the process is broken, which IMHO it really is.

**Comment by cartazio** [comment]
Help out or not.  Please shoot me an email on the process stuff. I’d love
some help and I balance my oss maintaining focus based upon user
communication and positive collaboration.

I will edit tickets so that I won’t rabbit hole on conflicts that don’t
matter.  This is to avoid pitfalls in my own personality where perceived
non technical conflicts distract me from doing an effective job supporting
users.  It reduces the number of imagined conflict panics down to zero.
That some people dislike that is a side effect. But cest lie vie

**Comment by cartazio** [comment]
To be clear : dev hasn’t been done on master. It’s more that Ryan’s merged
a number of prs. But I’m releasing some sooner than later so that I can
make sure fixes hit as wide a support range as possible.

**Comment by lehins** [comment]
I think that is the culprit
> dev hasn’t been done on master

Not only Ryan, but all other contributors expect dev to be happening on the default branch, which seems to be set to `master` for vector and that is really a defacto standard.

An alternative would be switching default repository branch to current major release branch `v12`, but that would be somewhat unorthodox, but at least it follow the current dev process. 

I don't think there is much more to say beyond what I already mentioned in my current and previous comments.
> Please shoot me an email on the process stuff

With regards to help, I'd be happy to submit a few PRs myself, there is at least three of the opened issues that I know of how to fix and would love to take care of. But considering that dev process is non-standard and there is no documented process established on on how to contribute to this repo I didn't even attempt working on those issues.

**Comment by cartazio** [comment]
Could you put together a checklist of current confusions we can go over and
I’ll do a bit of catchup to address or fix them ?

**Comment by cartazio** [comment]
ill take some time today to see what I need to do to catchup resync stuff

**Comment by cartazio** [comment]
Don’t worry about a theoretical checklist.  I’ll just go over what changes
have what pvp impact and sort them out.

I also should double check a feeble things

**Comment by cartazio** [comment]
Meta note : as an extrovert I definitely respond and prioritize prjects
where users / contributors communicate.  So I really do appreciate you
taking the time to articulate your frustrations. It helps me and thus helps
everyone

**Comment by Shimuuar** [comment]
Fixed in 0.12.1 release
